operator/pkg/ciliumidentity/reconciler.go

```diff
@@ -0,0 +1,444 @@
+
+	podName := podResourceKey(pod.Name, pod.Namespace).String()
+	prevCIDName, _ := r.cidUsageInPods.AssignCIDToPod(podName, cidName)
+
+	r.logger.Info("CID allocated for pod",
+		logfields.K8sPodName, fmt.Sprintf("%s/%s", pod.Namespace, pod.Name),
+		logfields.CIDName, cidName,
+		logfields.OldIdentity, prevCIDName,
+		logfields.Labels, k8sLabels)
+
+	if isNewCID {
+		r.queueOps.enqueueCIDReconciliation(cidResourceKey(cidName), 0)
+	}
+
+	return nil
+}
+
+func (r *reconciler) allocateCID(cidKey *key.GlobalIdentity) (string, bool, error) {
+	cidName, exists := r.desiredCIDState.LookupByKey(cidKey)
+	if exists {
+		return cidName, false, nil
+	}
+
+	storeCIDs, err := r.cidStore.ByIndex(k8s.ByKeyIndex, cidKey.GetKey())
+	if err != nil {
+		return "", false, err
+	}
+
+	// If CIDs that match labels are found in CID store but not in the desired cache,
+	// they need to be added to the desired cache and used instead of creating a new
+	// CID for these labels.
+	if len(storeCIDs) > 0 {
+		// Return the assignment from the CID store, otherwise allocates a new identity
+		cidName, err = r.handleStoreCIDMatch(storeCIDs)
+		if err != nil {
+			r.logger.Error("Failed to access CID store", logfields.Error, err)
+		} else {
+			return cidName, false, nil
+		}
+	}
+
+	allocatedID, err := r.idAllocator.AllocateRandom()
+	if err != nil {
+		return "", false, err
+	}
+
+	return allocatedID.String(), true, nil
+}
+
+func (r *reconciler) getRelevantLabelsForPod(pod *slim_corev1.Pod) (map[string]string, error) {
+	ns, err := r.getNamespace(pod.Namespace)
+	if err != nil {
+		return nil, err
+	}
+
+	_, labelsMap, _, err := k8s.GetPodMetadata(ns, pod)
+	if err != nil {
+		return nil, err
+	}
+
+	return labelsMap, nil
+}
+
+func (r *reconciler) getNamespace(namespace string) (*slim_corev1.Namespace, error) {
+	ns, exists, err := r.nsStore.GetByKey(resource.Key{Name: namespace})
+	if err != nil {
+		return nil, fmt.Errorf("unable to get namespace %q, error: %w", namespace, err)
+	}
+	if !exists {
+		return nil, fmt.Errorf("namespace %q not found in store", namespace)
+	}
+
+	return ns, nil
+}
+
+func (r *reconciler) handleStoreCIDMatch(storeCIDs []*cilium_api_v2.CiliumIdentity) (string, error) {
+	if len(storeCIDs) == 0 {
+		return "", fmt.Errorf("store CIDs list is empty")
+	}
+
+	var selectedCIDName string
+
+	// Deduplication: Reconcile all CID. The first will be added to the desired
+	// cache and the rest will be deleted, because they are not used.
+	for _, cid := range storeCIDs {
+		toDelete := true
+
+		if len(selectedCIDName) == 0 {
+			cidKey := key.GetCIDKeyFromLabels(cid.SecurityLabels, "")
+			if err := r.upsertDesiredState(cid.Name, cidKey); err != nil {
+				r.logger.Warn("Failed to add CID to cache",
+					logfields.CIDName, cid.Name,
+					logfields.Error, err)
+			} else {
+				toDelete = false
+				selectedCIDName = cid.Name
+			}
+		}
+
+		if toDelete {
+			r.queueOps.enqueueCIDReconciliation(cidResourceKey(cid.Name), 0)
+		}
+	}
```

Comment:
We are not deleting identities from the new controller, so is the call to `enqueueCIDReconciliation` (and the `toDelete` logic) necessary?

Reply:
refactored the method as we no longer need to reconcile all CIDs if operator managing CIDs is no longer deleting the CIDs.